snippet-extractor/python_src/snippet_extractor.py

```diff
@@ -0,0 +1,148 @@
+        self,
+        *,
+        primary: TextSplitter,
+        secondary: TextSplitter,
+        hard_chunk_size_limit: int,
+        length_function: Callable[[str], int]
+    ):
+        # setting chunk_size = hard_max_chunk is needed for using self._merge_splits
+        super().__init__(
+            chunk_size=hard_chunk_size_limit,
+            chunk_overlap=0,
+            length_function=length_function,
+        )
+        self._primary = primary
+        self._secondary = secondary
+        self._hard_chunk_size_limit = hard_chunk_size_limit
+
+    def split_text(self, text: str) -> List[str]:
+        main_splits = self._primary.split_text(text)
+
+        # remove snippets that are larger than hard_max_chunk
+        no_big_splits = []
+        for split in main_splits:
+            if self._length_function(split) > self._hard_chunk_size_limit:
+                secondary_splits = self._secondary.split_text(split)
+                no_big_splits.extend(secondary_splits)
+            else:
+                no_big_splits.append(split)
+
+        return self._merge_splits(no_big_splits, "\n")
+
+
+class SnippetExtractor(TextSplitterWithBigChunkSplitter):
+    def __init__(
+        self,
+        *,
+        language: str,
+        chunk_size: int,
+        hard_chunk_size_limit: int,
+        tokenizer,
+    ):
+        token_len = lambda s: len(tokenizer(s).input_ids)
+        super().__init__(
+            primary=NLTKTextSplitter(
+                language=language,
+                chunk_size=chunk_size,
+                chunk_overlap=0,
+                length_function=token_len,
+            ),
+            secondary=RecursiveCharacterTextSplitter(
+                chunk_size=chunk_size, chunk_overlap=0, length_function=token_len
+            ),
+            hard_chunk_size_limit=hard_chunk_size_limit,
+            length_function=token_len,
+        )
+
+def ok(value: any) -> dict:
+    return { 'Ok': value }
+def err(value: any) -> dict:
+    return { 'Err': str(value) }
+
+
+def run_stdio_client():
+    import sys
+    tokenizers = {}
+
+    # There is some bad interaction between stdin Binary I/O buffering code
+    # and the Unpacker with can lead to hangs. Using Raw I/O avoids this issue.
+    stdin = sys.stdin.buffer.raw
+    stdout = sys.stdout.buffer
+
+    # Disable buffer size safety check, we already checked it on the rust side
+    # and keeping that option is sync is another potential source of bugs.
+    unpacker = Unpacker(stdin, max_buffer_size = 0)
+    packer = Packer()
+
+    stdout.write(packer.pack("ready"));
+    stdout.flush()
+
+    for msg in unpacker:
+        result = None
+        try:
+            tag = msg['tag']
+            cmd = msg['cmd']
+            if tag == 'initialize_tokenizer':
+                tokenizers[cmd['name']] = PreTrainedTokenizerFast(tokenizer_file=cmd['path'])
+                result = ok(True)
+            elif tag == 'extract':
+                snippets = SnippetExtractor(
```

Comment:
This will read the model from the disk every time. On the other script we have, if I initialize it 100times it takes more than 1s.
We can use a cache like we are doing for the tokenizers so we take an hit only the first time.

Reply:
it doesn't only the tokenizer reads any model from disk, which we initialize beforhand

Reply:
I just realized you might not be speaking about the `punkt` model from the tokenizer.

But then which model?

The NLTK  split only creates a lambda, which might internally load some model, but if so does cache it internally AFIK as we don't create any state for it.

And the other splitters don't have any model.

Reply:
I was talking about punkt. I got a super high time because I was partially retraining the model to better understand legal abbreviations. That is super slow. but we should put that in a pickle file and use that as an asset.

Reply:
Currently we do load the punk tokenizer only once per process and keep it around as PreTrainedTokenizer, the `NLTK` send_tokens function also seems to load something from punk but also caches it.

So AFIK there is nothing which will do any disk access on every request.